File: crates/context/src/openai.rs
//! OpenAI-compatible wire format: the request-body writer and reply reader.
//!
//! PROVISIONAL (G4 discovery): the G3 freeze said "the exact JSON shape per
//! provider is serialization detail applied when the request body is
//! written" but gave that writer no home. It lives here because provider
//! quirks live only in `context` (§8.1); `core::execute_effect` calls these
//! and `ModelPort` moves the bytes without interpreting them.

use serde_json::{json, Value};

use crate::render::{ContentPart, Message, Role};

fn role_str(role: Role) -> &'static str {
    match role {
        Role::System => "system",
        Role::User => "user",
        Role::Assistant => "assistant",
    }
}

/// One neutral part → the OpenAI content-part union.
fn part_json(p: &ContentPart) -> Value {
    match p {
        ContentPart::Text { text } => json!({"type": "text", "text": text}),
        ContentPart::Image {
            media_type,
            data_base64,
        } => json!({"type": "image_url",
            "image_url": {"url": format!("data:{media_type};base64,{data_base64}")}}),
        ContentPart::Audio {
            media_type,
            data_base64,
        } => json!({"type": "input_audio", "input_audio": {"data": data_base64,
            "format": media_type.split('/').next_back().unwrap_or("wav")}}),
        ContentPart::File {
            name,
            media_type,
            data_base64,
        } => json!({"type": "file", "file": {"filename": name,
            "file_data": format!("data:{media_type};base64,{data_base64}")}}),
    }
}

/// The `/v1/chat/completions` request body. Text-only messages collapse to
/// plain string content (widest local-server compatibility); mixed content
/// uses the array form. `stream: false` — streaming is core-driven chaining
/// over completed replies (ADR-002), not SSE.
pub fn openai_request_body(messages: &[Message], model: &str) -> String {
    let msgs: Vec<Value> = messages
        .iter()
        .map(|m| {
            let all_text = m
                .content
                .iter()
                .all(|p| matches!(p, ContentPart::Text { .. }));
            let content: Value = if all_text {
                m.content
                    .iter()
                    .map(|p| match p {
                        ContentPart::Text { text } => text.as_str(),
                        _ => unreachable!(),
                    })
                    .collect::<Vec<_>>()
                    .join("")
                    .into()
            } else {
                Value::Array(m.content.iter().map(part_json).collect())
            };
            json!({"role": role_str(m.role), "content": content})
        })
        .collect();
    json!({"model": model, "stream": false, "messages": msgs}).to_string()
}

/// Extract the assistant text from a chat-completion reply body. `None` when
/// the body isn't a recognizable completion — the caller surfaces that as a
/// typed model error, never a fake reply.
pub fn openai_reply_text(body_json: &str) -> Option<String> {
    let v: Value = serde_json::from_str(body_json).ok()?;
    let message = v.get("choices")?.get(0)?.get("message")?;
    let text = match message.get("content") {
        Some(Value::String(s)) => Some(s.clone()),
        Some(Value::Array(parts)) => Some(
            parts
                .iter()
                .filter_map(|p| p.get("text").and_then(Value::as_str))
                .collect::<Vec<_>>()
                .join(""),
        ),
        _ => None,
    };
    match text.filter(|t| !t.trim().is_empty()) {
        Some(text) => Some(text),
        // A reply with NO content but native `tool_calls`. This build asks for
        // calls as text (the Python's layout rule is a text rule), but a
        // provider may answer in its own tool-call shape anyway — omlx does,
        // for a prompt whose affordances mention tools. Reading that as "no
        // reply" would discard a call the model really made and stall the turn
        // on "unrecognizable completion body". Rendered into the ONE call
        // syntax the parser reads, so nothing downstream learns a second one.
        None => native_calls(message),
    }
}

/// `tool_calls` as the text this build's parser already understands: one call
/// per line, which is also the layout rule's "these run in order".
fn native_calls(message: &Value) -> Option<String> {
    let calls = message.get("tool_calls")?.as_array()?;
    let lines: Vec<String> = calls
        .iter()
        .filter_map(|call| {
            let f = call.get("function")?;
            // Providers namespace the name (`tools:list_agents`); the toolbox
            // knows the bare one, and an unknown tool is refused in words.
            let name = f.get("name")?.as_str()?.rsplit(':').next()?;
            let args = match f.get("arguments") {
                Some(Value::String(s)) if !s.trim().is_empty() => s.trim().to_string(),
                Some(Value::Object(o)) => Value::Object(o.clone()).to_string(),
                _ => "{}".to_string(),
            };
            Some(format!("{name}({args})"))
        })
        .collect();
    match lines.is_empty() {
        true => None,
        false => Some(lines.join("\n")),
    }
}

File: packages/context/src/openai.js
/**
 * The OpenAI chat-completions protocol — and therefore nearly the whole
 * catalogue, since llama.cpp, LM Studio, vLLM, OpenRouter and DeepSeek all
 * answer on it.
 *
 * THE ONE LINE THAT BRICKS A SESSION. An assistant turn carrying only
 * reasoning, or only tool calls, must serialise `content` as `""` and NEVER
 * `null`. The message sits durably in the session log, so a single null is not
 * one bad turn — it is every later turn of that session, refused by an API
 * that will not say which message it choked on.
 *
 * REASONING PASSBACK IS CONDITIONAL ON `tools`. With tools present, DeepSeek
 * requires the intermediate `reasoning_content` back on every subsequent turn,
 * INCLUDING turns where the model called nothing, or answers 400. With no
 * tools it wants none of it — and `ports.js` keeps reasoning out of history by
 * default for exactly that case. Both polarities are here, and the condition
 * is the tool list this request carries.
 * @module
 */

import { messagesOf } from './wire.js'
import { finishFrom, ownReplay } from './provider.js'
import { at, count, list, readBody, str } from './read.js'
import { IMAGE_RULES } from './image.js'

/** @typedef {import('./provider.js').ProviderAdapter} ProviderAdapter */
/** @typedef {import('./provider.js').ProviderReply} ProviderReply */
/** @typedef {import('./provider.js').Exchange} Exchange */
/** @typedef {import('./provider.js').ToolSpec} ToolSpec */
/** @typedef {import('./wire.js').Message} Message */
/** @typedef {import('./types.js').Part} Part */

const PROVIDER = 'openai'

/** @type {Record<string, import('@harness/kernel/ports.js').FinishReason>} */
const FINISH = {
  stop: 'stop',
  tool_calls: 'tool_calls',
  length: 'length',
  content_filter: 'content_filter',
}

/** @type {ProviderAdapter} */
export const openaiAdapter = {
  provider: PROVIDER,
  images: IMAGE_RULES.openai,
  buildRequest,
  parseResponse,
}

/**
 * The `/v1/chat/completions` body: the paper's system message, every
 * replayable assistant turn with its results, then the paper's user message.
 * `temperature` is OMITTED when the agent file named none — an absent key means
 * the endpoint's own default, and a number we invented would override a server
 * setting nobody asked us to touch.
 * @param {import('./types.js').Document} doc
 * @param {import('./card.js').ModelCard} card
 * @param {ToolSpec[]} tools
 * @param {import('./provider.js').RequestOpts} [opts]
 * @returns {Record<string, unknown>}
 */
function buildRequest(doc, card, tools, opts = {}) {
  const paper = messagesOf(doc, card).map(wireMessage)
  const replayed = (opts.replay ?? []).flatMap((t) => replayMessages(t, tools.length > 0))
  /** @type {Record<string, unknown>} */
  const body = {
    model: card.model,
    stream: opts.stream === true,
    messages: [paper[0], ...replayed, paper[1]],
  }
  if (tools.length > 0) {
    body['tools'] = tools.map((t) => ({
      type: 'function',
      function: { name: t.name, description: t.description, parameters: t.parameters },
    }))
  }
  if (typeof opts.temperature === 'number') body['temperature'] = opts.temperature
  return body
}

/**
 * One earlier assistant turn, plus one message per tool result. The result
 * messages are not optional: an assistant message carrying `tool_calls` whose
 * results never arrive is refused.
 * @param {Exchange} turn @param {boolean} withTools
 * @returns {Array<Record<string, unknown>>}
 */
function replayMessages(turn, withTools) {
  const state = ownReplay(turn, PROVIDER)
  // `str` and not `turn.text`: the turn came back out of the session log, and
  // a null that got in there once must not be written back to the wire.
  /** @type {Record<string, unknown>} */
  const message = { role: 'assistant', content: str(turn.text) }
  if (turn.calls.length > 0) {
    message['tool_calls'] = turn.calls.map((c) => ({
      id: c.id,
      type: 'function',
      function: { name: c.tool, arguments: c.args },
    }))
  }
  const reasoning = str(at(state, 'reasoning'))
  if (withTools && reasoning) message['reasoning_content'] = reasoning
  return [message, ...turn.results.map((r) => ({ role: 'tool', tool_call_id: r.id, content: r.output }))]
}

/**
 * A rendered message on the wire. Text-only content collapses to a plain
 * string — the widest local-server compatibility; mixed content uses the array
 * form. `content` is `''` and never null in both branches.
 * @param {Message|undefined} m @returns {Record<string, unknown>}
 */
function wireMessage(m) {
  const parts = m?.content ?? []
  const text = parts.every((p) => p.type === 'text')
  return {
    role: m?.role ?? 'user',
    content: text ? parts.map((p) => (p.type === 'text' ? p.text : '')).join('') : parts.map(partJson),
  }
}

/** One neutral part as the OpenAI content-part union. @param {Part} p */
function partJson(p) {
  switch (p.type) {
    case 'text':
      return { type: 'text', text: p.text }
    case 'image':
      return { type: 'image_url', image_url: { url: `data:${p.mediaType};base64,${p.dataBase64}` } }
    case 'audio':
      return { type: 'input_audio', input_audio: { data: p.dataBase64, format: p.mediaType.split('/').pop() ?? 'wav' } }
    case 'file':
      return { type: 'file', file: { filename: p.name, file_data: `data:${p.mediaType};base64,${p.dataBase64}` } }
  }
}

/** @param {unknown} body @returns {ProviderReply} */
function parseResponse(body) {
  const raw = readBody(body, PROVIDER)
  const message = at(list(at(raw, 'choices'))[0], 'message')
  const reasoning = str(at(message, 'reasoning_content')) || str(at(message, 'reasoning'))
  return {
    text: textOf(at(message, 'content')),
    reasoning,
    calls: callsOf(list(at(message, 'tool_calls'))),
    finish: finishFrom(FINISH, at(list(at(raw, 'choices'))[0], 'finish_reason')),
    usage: usageOf(at(raw, 'usage')),
    raw,
    provider: PROVIDER,
    replayState: { reasoning },
  }
}

/** Content as text, whichever of the two shapes the server used. @param {unknown} content */
function textOf(content) {
  if (typeof content === 'string') return content
  return list(content)
    .map((p) => str(at(p, 'text')))
    .join('')
}

/**
 * Native tool calls. The name is taken BARE: providers namespace it
 * (`tools:list_agents`) and the toolbox knows only the last segment.
 * @param {unknown[]} calls
 */
function callsOf(calls) {
  return calls.flatMap((c) => {
    const fn = at(c, 'function')
    const name = str(at(fn, 'name')).split(':').pop() ?? ''
    if (!name) return []
    const args = at(fn, 'arguments')
    return [{ id: str(at(c, 'id')), tool: name, args: typeof args === 'string' ? args : JSON.stringify(args ?? {}) }]
  })
}

/**
 * The accounting block, normalised. `cached_tokens` is a SUBSET of
 * `prompt_tokens` here, and DeepSeek folds its cache hits into that same
 * total, so the cached count is subtracted back out of the input rather than
 * counted twice. `reasoning_tokens` is carried and never added: it is already
 * inside `completion_tokens`.
 * @param {unknown} usage @returns {import('./provider.js').ProviderUsage|null}
 */
function usageOf(usage) {
  const input = count(at(usage, 'prompt_tokens'))
  const output = count(at(usage, 'completion_tokens'))
  if (input === null && output === null) return null
  const cached =
    count(at(at(usage, 'prompt_tokens_details'), 'cached_tokens')) ?? count(at(usage, 'prompt_cache_hit_tokens'))
  return {
    inputTokens: Math.max(0, (input ?? 0) - (cached ?? 0)),
    outputTokens: output ?? 0,
    cachedInputTokens: cached,
    reasoningTokens: count(at(at(usage, 'completion_tokens_details'), 'reasoning_tokens')),
  }
}
